server/src/pages/shot/text-tool.js

```diff
@@ -0,0 +1,205 @@
+const React = require("react");
+const PropTypes = require("prop-types");
+const { Localized } = require("fluent-react/compat");
+const sendEvent = require("../../browser-send-event.js");
+const { Selection } = require("../../../shared/selection");
+const { ColorPicker } = require("./color-picker");
+
+// Text input height and padding is determined by text size
+// See frame.scss mixin type-box($size)
+// TEXT_INPUT_PADDING_TOP is top padding and
+// TEXT_INPUT_PADDING_LEFT is multiplier used to calculate left padding
+const TEXT_INPUT_PADDING_TOP = 4;
+const TEXT_INPUT_PADDING_LEFT = 0.4;
+const TEXT_INPUT_HEIGHT = 1.3;
+
+// Padding used by draggable div
+const DRAG_PADDING = 10;
+const FONT_STYLE = "sans-serif";
+const FONT_WEIGHT = 900;
+const INIT_FONT_SIZE = 36;
+
+exports.TextTool = class TextTool extends React.Component {
+  constructor(props) {
+    super(props);
+    this.textInput = React.createRef();
+
+    this.canvasCssWidth = props.canvasCssWidth;
+    this.canvasCssHeight = props.canvasCssHeight;
+
+    const INIT_LEFT = (this.canvasCssWidth / 2) - INIT_FONT_SIZE * TEXT_INPUT_PADDING_LEFT;
+    const INIT_TOP = (this.canvasCssHeight / 2) - ((INIT_FONT_SIZE * TEXT_INPUT_HEIGHT) / 2);
+
+    this.state = {
+      left: INIT_LEFT,
+      top: INIT_TOP,
+      color: this.props.color,
+      colorName: this.props.colorName,
+      textSize: "large-text"
+    };
+  }
+
+  componentDidMount() {
+    this.textInput.current.focus();
+
+    if (this.props.toolbarOverrideCallback) {
+      this.props.toolbarOverrideCallback();
+    }
+  }
+
+  componentDidUpdate(oldProps, oldState) {
+    if (oldState.textSize !== this.state.textSize) {
+       this.props.toolbarOverrideCallback();
+    }
+  }
+
+  componentWillUnmount() {
+    if (this.props.toolbarOverrideCallback) {
+      this.props.toolbarOverrideCallback();
+    }
+  }
+
+  render() {
+    const textToolBox = this.renderTextBox();
+    return <div
+      className={`text-tool-container centered`}
+      onDragOver={this.onDragOver.bind(this)}
+      onDrop={this.onDrop.bind(this)}
+      style={{height: this.props.baseCanvas.style.height, width: this.props.baseCanvas.style.width}}>
+      {textToolBox}
+    </div>;
+  }
+
+  renderTextBox() {
+    const dragDivTopPx = `${this.state.top - DRAG_PADDING}px`;
+    const dragDivLeftPx = `${this.state.left - DRAG_PADDING}px`;
+    const dragDivPadding = `${DRAG_PADDING}px`;
+
+    const dragDivStyles = {
+      position: "absolute",
+      top: dragDivTopPx,
+      left: dragDivLeftPx,
+      padding: dragDivPadding,
+      cursor: "move"
+    };
+
+    return [
+      <div draggable="true" key="drag" onDragStart={this.onDragStart.bind(this)} style={dragDivStyles}>
+        <div id="text-input" ref={this.textInput} contentEditable="true" key="text"
+           className={`${this.state.textSize} ${this.state.colorName} text`} >
+        </div>
+      </div>
```

Comment:
We should change the initial position of this.  If the user has a very long shot, and he scrolls to the spot where he wants to add some text, this will cause a jump away from the user's desired position.

Reply:
@chenba Is the ask here to change initial position to middle of viewport, cannot replicate the jump away from desired position part when user scrolls on a very long shot

Reply:
Updated to change initial position to middle of viewport for a very long shot